Mathlib/RingTheory/DedekindDomain/Different.lean

```diff
@@ -0,0 +1,287 @@
+-/
+
+universe u
+
+attribute [local instance] FractionRing.liftAlgebra FractionRing.isScalarTower_liftAlgebra
+
+variable (A K) {L : Type u} {B} [CommRing A] [Field K] [CommRing B] [Field L]
+variable [Algebra A K] [Algebra B L] [Algebra A B] [Algebra K L] [Algebra A L]
+variable [IsScalarTower A K L] [IsScalarTower A B L]
+variable [IsDomain A] [IsDomain B]
+variable [IsFractionRing A K] [IsIntegralClosure B A L] [IsFractionRing B L]
+variable [FiniteDimensional K L] [IsSeparable K L]
+
+open nonZeroDivisors IsLocalization Matrix Algebra
+
+/-- Under the AKLB setting, `Iᵛ := traceFormDualSubmodule A K (I : Submodule B L)` is the
+`Submodule B L` such that `x ∈ Iᵛ ↔ ∀ y ∈ I, Tr(x, y) ∈ A` -/
+def traceFormDualSubmodule (I : Submodule B L) : Submodule B L where
+  carrier := { s | ∀ a ∈ I, traceForm K L s a ∈ (algebraMap A K).range }
+  add_mem' hx hy a ha :=
+    BilinForm.add_left (B := traceForm K L) _ _ _ ▸ add_mem (hx a ha) (hy a ha)
+  zero_mem' a _ := BilinForm.zero_left (B := traceForm K L) a ▸ zero_mem _
+  smul_mem' c x hx a ha := by
+    rw [traceForm_apply, smul_mul_assoc, mul_comm, ← smul_mul_assoc, mul_comm]
+    exact hx _ (Submodule.smul_mem _ c ha)
+
+variable {A K}
+
+local notation:max I:max "ᵛ" => traceFormDualSubmodule A K I
+
+lemma mem_traceFormDualSubmodule {I : Submodule B L} {x} :
+    x ∈ Iᵛ ↔ ∀ a ∈ I, traceForm K L x a ∈ (algebraMap A K).range :=
+  Iff.rfl
+
+lemma le_traceFormDualSubmodule_iff_map_le_one {I J : Submodule B L} :
+    I ≤ Jᵛ ↔ ((I * J : Submodule B L).restrictScalars A).map
+      ((trace K L).restrictScalars A) ≤ 1 := by
+  rw [Submodule.map_le_iff_le_comap, Submodule.restrictScalars_mul, Submodule.mul_le]
+  simp [SetLike.le_def, traceFormDualSubmodule]
+
+lemma le_traceFormDualSubmodule_mul_iff {I J J' : Submodule B L} :
+    I ≤ (J * J')ᵛ ↔ I * J ≤ J'ᵛ := by
+  simp_rw [le_traceFormDualSubmodule_iff_map_le_one, mul_assoc]
+
+lemma le_traceFormDualSubmodule {I J : Submodule B L} :
+    I ≤ Jᵛ ↔ I * J ≤ 1ᵛ := by
+  rw [← le_traceFormDualSubmodule_mul_iff, mul_one]
+
+lemma le_traceFormDualSubmodule_comm {I J : Submodule B L} :
+    I ≤ Jᵛ ↔ J ≤ Iᵛ := by rw [le_traceFormDualSubmodule, mul_comm, ← le_traceFormDualSubmodule]
+
+lemma le_traceFormDualSubmodule_traceFormDualSubmodule {I : Submodule B L} :
+    I ≤ Iᵛᵛ := le_traceFormDualSubmodule_comm.mpr le_rfl
+
+variable [IsIntegrallyClosed A]
+
+lemma mem_traceFormDualSubmodule_iff_isIntegral {I : Submodule B L} {x} :
+    x ∈ Iᵛ ↔ ∀ a ∈ I, IsIntegral A (traceForm K L x a) :=
+  forall₂_congr (fun _ _ ↦ IsIntegrallyClosed.isIntegral_iff.symm)
+
+/-- If `b` is an `A`-integral basis of `L` with discriminant `b`, then `d • a * x` is integral over
+  `A` for all `a ∈ I` and `x ∈ Iᵛ`. -/
+lemma isIntegral_discr_mul_of_mem_traceFormDualSubmodule
+    (I : Submodule B L) {ι} [DecidableEq ι] [Fintype ι]
+    (b : Basis ι K L) (hb : ∀ i, IsIntegral A (b i))
+    (a x : L) (ha : a ∈ I) (hx : x ∈ Iᵛ) :
+    IsIntegral A ((discr K b) • a * x) := by
```

Comment:
```suggestion
lemma isIntegral_discr_mul_of_mem_traceFormDualSubmodule
    (I : Submodule B L) {ι} [DecidableEq ι] [Fintype ι]
    {b : Basis ι K L} (hb : ∀ i, IsIntegral A (b i))
    {a x : L} (ha : a ∈ I) (hx : x ∈ Iᵛ) :
    IsIntegral A ((discr K b) • a * x) := by
```
Maybe the name can be changed to allow dot  notation with `Basis`, but I am not sure.

Reply:
I don't think so, as `b : Basis` is implicit argument, and it is not the main focus of the lemma.